From f02d64ad45acc2282e0982962fc83dfd433b286e Mon Sep 17 00:00:00 2001
From: Jonathan DE CESCO <jonathan.de.cesco@intel.com>
Date: Wed, 5 Sep 2012 17:33:54 +0200
Subject: [PORT FROM MAIN] wl12xx: Log the packets that caused WiFi to wake up the platform

BZ: 42116

Finding why WiFi woke up the platform can be difficult involving using Wireshark or
Omnipeek network sniffers.
This patch will log the N packets (N=2) that is received after the WiFi chip
woke up the platform. With N=2, we will hopefully log the packet that
caused the wakeup, since typically we have an ARP packet and an IP packet.
Enabled by default only on 'eng' builds.

Change-Id: I83b5945b344ff720d9b8d85178b95258d981c1bf
Orig-Change-Id: I098c5424db025ef251be097a91c76f79e1437e9f
Signed-off-by: Frode Isaksen <frodex.isaksen@intel.com>
Signed-off-by: Jonathan DE CESCO <jonathan.de.cesco@intel.com>
Reviewed-on: http://android.intel.com:8080/64576
Tested-by: Champciaux, Nicolas <nicolas.champciaux@intel.com>
Reviewed-by: buildbot <buildbot@intel.com>
Tested-by: buildbot <buildbot@intel.com>
Reviewed-on: http://android.intel.com:8080/77987
Reviewed-by: Li, XiaojingX <xiaojingx.li@intel.com>
Tested-by: Li, XiaojingX <xiaojingx.li@intel.com>
---
 wl12xx-compat/drivers/net/wireless/wl12xx/debug.h  |    1 +
 wl12xx-compat/drivers/net/wireless/wl12xx/main.c   |    3 +++
 wl12xx-compat/drivers/net/wireless/wl12xx/rx.c     |    5 +++++
 wl12xx-compat/drivers/net/wireless/wl12xx/wl12xx.h |    3 +++
 4 files changed, 12 insertions(+), 0 deletions(-)

diff --git a/wl12xx-compat/drivers/net/wireless/wl12xx/debug.h b/wl12xx-compat/drivers/net/wireless/wl12xx/debug.h
index ec0fdc2..e62fc3c 100644
--- a/wl12xx-compat/drivers/net/wireless/wl12xx/debug.h
+++ b/wl12xx-compat/drivers/net/wireless/wl12xx/debug.h
@@ -52,6 +52,7 @@ enum {
 	DEBUG_ADHOC     = BIT(16),
 	DEBUG_AP	= BIT(17),
 	DEBUG_PROBE	= BIT(18),
+	DEBUG_WAKE	= BIT(19),
 	DEBUG_MASTER	= (DEBUG_ADHOC | DEBUG_AP),
 	DEBUG_ALL	= ~0,
 };
diff --git a/wl12xx-compat/drivers/net/wireless/wl12xx/main.c b/wl12xx-compat/drivers/net/wireless/wl12xx/main.c
index a968998..5ba38fd 100644
--- a/wl12xx-compat/drivers/net/wireless/wl12xx/main.c
+++ b/wl12xx-compat/drivers/net/wireless/wl12xx/main.c
@@ -2342,6 +2342,9 @@ static int wl1271_op_resume(struct ieee80211_hw *hw)
 
 		/* don't talk to the HW if recovery is pending */
 		if (!pending_recovery) {
+			if (DEBUG_WAKE & wl12xx_debug_level) {
+				wl->log_wake_pkts = 2; /* Log 2 packets after wake, typically ARP + IP */
+			}
 			ret = wl12xx_irq_locked(wl);
 			if (ret)
 				wl12xx_queue_recovery_work(wl);
diff --git a/wl12xx-compat/drivers/net/wireless/wl12xx/rx.c b/wl12xx-compat/drivers/net/wireless/wl12xx/rx.c
index 3dd569c..79172d3 100644
--- a/wl12xx-compat/drivers/net/wireless/wl12xx/rx.c
+++ b/wl12xx-compat/drivers/net/wireless/wl12xx/rx.c
@@ -179,6 +179,11 @@ static int wl1271_rx_handle_data(struct wl1271 *wl, u8 *data, u32 length,
 
 	skb_trim(skb, skb->len - desc->pad_len);
 
+	if (!beacon && (wl->log_wake_pkts > 0)) {
+		wl1271_dump_ascii(DEBUG_WAKE, "wake: ", skb->data, min(skb->len, 128));
+		--wl->log_wake_pkts;
+	}
+
 	skb_queue_tail(&wl->deferred_rx_queue, skb);
 	queue_work(wl->freezable_wq, &wl->netstack_work);
 
diff --git a/wl12xx-compat/drivers/net/wireless/wl12xx/wl12xx.h b/wl12xx-compat/drivers/net/wireless/wl12xx/wl12xx.h
index c42562a..52b1df4 100644
--- a/wl12xx-compat/drivers/net/wireless/wl12xx/wl12xx.h
+++ b/wl12xx-compat/drivers/net/wireless/wl12xx/wl12xx.h
@@ -552,6 +552,9 @@ struct wl1271 {
 	/* RX Data filter rule status - enabled/disabled */
 	bool rx_data_filters_status[WL1271_MAX_RX_FILTERS];
 
+	/* Number of packets to log after WiFi wake */
+	int log_wake_pkts;
+
 	/* AP's peers */
 	struct list_head peers_list;
 
-- 
1.7.4.1

